.github/workflows/go-tests.yml

```diff
@@ -498,20 +498,36 @@ jobs:
     steps:
       - name: evaluate upstream job results
         run: |
+          NEEDS_JSON='${{ toJSON(needs) }}'
+          
+          # if failure (not cancelled), notify Slack
+          if echo "${NEEDS_JSON}" | grep -E -i '\"result\": \"(failure)\"'; then
+            printf "Tests failed, notifying Slack"
+            echo "FAILED_TESTS=true" >> $GITHUB_ENV
+          fi
+          
           # exit 1 if failure or cancelled result for any upstream job
-          if printf '${{ toJSON(needs) }}' | grep -E -i '\"result\": \"(failure|cancelled)\"'; then
+          # this ensures that we fail the PR check regardless of cancellation, rather than skip-passing it
+          # see https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution#overview
+          if echo "${NEEDS_JSON}" | grep -E -i '\"result\": \"(failure|cancelled)\"'; then
             printf "Tests failed or workflow cancelled:\n\n${{ toJSON(needs) }}"
             exit 1
           fi
+      - name: Set Slack commit message summary
+        # avoid duplicating branch filter logic below and enable easier message content debugging
+        # always() ensures this runs even if the test eval step exits 1
+        if: always()
+        run: echo "COMMIT_MESSAGE_SUMMARY=$(echo ${{ github.event.head_commit.message }} | head -n 1)" >> $GITHUB_ENV
```

Comment:
Could this step be part of the `Notify Slack` step below? I'm not quite sure what `# avoid duplicating branch filter logic below` means here.

Reply:
> Could this step be part of the Notify Slack step below?

I might be missing an option for that in GHA, but I _believe_ it has to be a separate step since it involves a custom shell script, and the next step is defined by `slackapi/slack-github-action` 

> I'm not quite sure what # avoid duplicating branch filter logic below means here.

I was just trying to call out that `if: always()` could have technically been `if: ${{ always() && env.FAILED_TESTS == 'true' && (github.ref_name == 'main' || startsWith(github.ref_name, 'release/')) }}` like below, but this is opting for simplicity. The `always()` part is necessary regardless of whether we filter it further.

Reply:
Ah I see! Makes sense to me